src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -330,6 +428,55 @@ class TestRequestRateManager : public TestLoadManagerBase,
  private:
   bool use_mock_infer_;
 
+  void TestCustomDataSetup(std::vector<ModelTensor>& tensors)
+  {
+    std::shared_ptr<MockDataLoader> mdl{
+        std::make_shared<MockDataLoader>(params_.batch_size)};
+
+    std::shared_ptr<MockModelParser> mmp{
+        std::make_shared<MockModelParser>(on_sequence_model_, false)};
+    mmp->inputs_ = std::make_shared<ModelTensorMap>();
+    for (auto t : tensors) {
+      (*mmp->inputs_)[t.name_] = t;
+    }
+
+    infer_data_manager_ =
+        MockInferDataManagerFactory::CreateMockInferDataManager(
+            params_.batch_size, params_.shared_memory_type,
+            params_.output_shm_size, mmp, factory_, mdl);
+
+    parser_ = mmp;
+    data_loader_ = mdl;
+    using_json_data_ = true;
+    execute_ = true;
+    max_threads_ = 1;
+  }
+
+  cb::Error TestCustomDataRunThread()
+  {
+    RateSchedulePtr_t schedule = std::make_shared<RateSchedule>();
+    schedule->intervals = NanoIntervals{milliseconds(4), milliseconds(8),
+                                        milliseconds(12), milliseconds(16)};
+    schedule->duration = nanoseconds{16000000};
```

Comment:
I am not a fan of these nondeterministic tests. Is there any way to skip the schedule and just directly call the function that has logic that's being tested repeatedly?

Like how I did here in my testing story:

https://github.com/triton-inference-server/client/blob/main/src/c%2B%2B/perf_analyzer/test_infer_context.cc#L104-L109

My concern is that this is a lot of test infrastructure that's hard to understand and therefore prone to bugs and friction for scaling/modifying.

Reply:
There is a lot of setup that happens. All of the params passed in to the manager go to specific places. The manager calls a few init functions, which among other things populates the shared memory.  I could manually do all of those things, but it would fork the code too much from what is actually happening, which is fragile and prone to masking errors.

I think a better path would be to keep this mostly the same, but to expose SendInferRequest() in a (mock) worker using gmock stuff. Then we can call that function 4 times. I would prefer to push that to a follow-on story so that all of our tests are fixed, not just this one.